include/dmlc/intrusive_ptr.h

```diff
@@ -0,0 +1,259 @@
+/*!
+ * Copyright (c) by Contributors 2020
+ * \file intrusive_ptr.h
+ * \brief Implementation of Intrusive Ptr.
+ */
+#ifndef DMLC_INTRUSIVE_PTR_H_
+#define DMLC_INTRUSIVE_PTR_H_
+
+#include <atomic>
+#include <cinttypes>
+#include <functional>
+
+namespace dmlc {
+/*!
+ * \brief Helper class for embedding reference counting into client objects.
+ */
+class IntrusivePtrCell {
+  std::atomic<int32_t> count_;
+  template <typename T> friend class IntrusivePtr;
+
+  std::int32_t Inc() noexcept { return ++count_; }
+  std::int32_t Dec() noexcept { return --count_; }
+  bool IsZero() const { return count_ == 0; }
+
+ public:
+  IntrusivePtrCell() noexcept : count_{0} {}
+  int32_t Count() const { return count_; }
+};
+
+/*!
+ * \brief User defined function for returing embedded reference count.
+ */
+template <typename T> IntrusivePtrCell &IntrusivePtrRefCount(T const *ptr) noexcept;
+
+/*!
+ * \brief Implementation of Intrusive Pointer.
+ *
+ * \code
+ *   struct RefCountedData {
+ *     mutable class IntrusivePtrCell ref;
+ *     float data { 0 };
+ *
+ *     friend IntrusivePtrCell &
+ *     IntrusivePtrRefCount(RefCountedData const *t) noexcept {
+ *       return t->ref;
+ *     }
+ *   };
+ *
+ *   auto p = make_intrusive<RefCountedData>();
+ *
+ * \endcode
+ */
+template <typename T> class IntrusivePtr {
+ private:
+  void Incref(T *ptr) {
+    if (ptr) {
+      IntrusivePtrRefCount(ptr).Inc();
+    }
+  }
+  void Decref(T *ptr) {
+    if (ptr) {
+      if (IntrusivePtrRefCount(ptr).Dec() == 0) {
+        delete ptr;
+      }
+    }
+  }
+
+  /*!
+   * \brief Contruct an IntrusivePtr.
+   *
+   * \param p Raw pointer to object
+   */
+  explicit IntrusivePtr(T *p) : ptr_{p} {
```

Comment:
Disallow constructor from raw ptr as it is unsafe, at least this should be made as a private function, and use something like IntrinsivePtr<T>::FromRaw

Reply:
This is private.